---
title: "3D'omics | Half-volume reaction and Celero adaptor test"
subtitle: "Manuscript in prep"
author:
  - Antton Alberdi^[University of Copenhagen, antton.alberdi@sund.ku.dk]
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
url: https://3d-omics.github.io/ANMI006_volume_adaptor_test
description: |
  Data analysis code for data analysis of the half-volume reaction and Celero adaptor test

link-citations: yes
github-repo: 3d-omics/ANMI006_volume_adaptor_test
---

```{r knitr_opts, echo=FALSE}
knitr::opts_chunk$set(
    class.source = "script-source",
    class.output = "script-output",
    comment = NA)
```

# Introduction

This webbook contains all the code used for the study on the effect of probiotics on Salmonella infection in broiler chicken.

## Prepare the R environment

### Environment

To reproduce all the analyses locally, clone this repository in your computer using:

```
RStudio > New Project > Version Control > Git
```

And indicating the following git repository:

> https://github.com/3d-omics/AMAC005_salmonella_trial.git

Once the R project has been created, follow the instructions and code chunks shown in this webbook.

### Libraries

The following R packages are required for the data analysis.

```{r load_libraries, warning=FALSE, comments="", message=FALSE}
# Base
library(R.utils)
library(knitr)
library(tidyverse)
library(devtools)
library(tinytable)
library(broom)
library(broom.mixed)
library(rairtable)

# For tree handling
library(ape)
library(phyloseq)
library(phytools)

# For plotting
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggnewscale)
library(gridExtra)
library(ggtreeExtra)
library(ggtree)
library(ggh4x)

# For statistics
library(spaa)
library(vegan)
library(Rtsne)
library(geiger)
library(hilldiv2)
library(distillR)
library(ANCOMBC)
library(lme4)
library(Hmsc)
```

<!--chapter:end:index.Rmd-->

# Data preparation

### Sample metadata

```{r sample_metadata, message=FALSE}
sample_metadata <- airtable("4-MSE-Info", "appKakM1bnKSekwuW") %>% #get base ID from Airtable browser URL
  read_airtable(., fields = c("ID","LabBatch_text","IntestinalSection","SampleType","Xcoord","Ycoord","size","cryosection_text","buffer_text"), id_to_col = TRUE) %>% #get 3 columns from MAGs table
  filter(LabBatch_text %in% c("MSEB0006","MSEB0009")) %>%
  rename(batch=LabBatch_text,microsample=ID,section=IntestinalSection,type=SampleType,cryosection=cryosection_text,buffer=buffer_text) %>%
  select(microsample,section,type,batch,cryosection,buffer,Xcoord,Ycoord,size) %>%
  unnest(c(section, Xcoord, Ycoord, size)) %>%
  mutate(protocol = case_when(
        microsample %in% c("M300682","M300683","M300684","M300685","M300686","M300687","M300688","M300689","M300690","M300691","M300692","M300693","M300694","M300695","M300696","M300697") ~ "half_ulv2",
        microsample %in% c("M300650","M300651","M300652","M300653","M300654","M300655","M300656","M300657","M300658","M300659","M300660","M300661","M300662","M300663","M300664","M300665","M300666","M300667","M300668","M300669","M300670","M300671","M300672","M300673","M300674","M300675","M300676","M300677","M300678","M300679","M300680","M300681") ~ "full_celero",
        TRUE ~ "full_ulv2")) %>%
  arrange(microsample)
```

## Count data

```{r load_count, message=FALSE}
read_counts <- read_tsv("data/MSEB0006_read_counts.tsv") %>%
  left_join(read_tsv("data/MSEB0009_read_counts.tsv"),by="sequence_id") %>%
  rename(genome = 1) %>%
  pivot_longer(!genome, names_to = "data", values_to="counts") %>%
  mutate(sample = substr(data, 1, 7)) %>%
  group_by(genome,sample) %>%
  summarise(counts=sum(counts), .groups="drop") %>%
  pivot_wider(names_from="sample", values_from="counts")
```

### Base hit table
This is the document containing the number of nucleotide bases have been covered by at least one read in each sample and MAG. This information is used to calculate MAG coverage values.

```{r load_hits, message=FALSE}
basehits <- read_tsv("data/MSEB0006_covered_bases.tsv") %>%
  left_join(read_tsv("data/MSEB0009_covered_bases.tsv"),by="sequence_id") %>%
  rename(genome = 1) %>%
  pivot_longer(!genome, names_to = "data", values_to="counts") %>%
  mutate(sample = substr(data, 1, 7)) %>%
  group_by(genome,sample) %>%
  summarise(counts=sum(counts), .groups="drop") %>%
  pivot_wider(names_from="sample", values_from="counts")
```

### Genome metadata
Relevant metadata of genomes is fetched from 2-3 files and merged into one genome metadata object for downstream analyses.

#### Taxonomy
This is the raw taxonomy table generated by GTDBtk, which is simplified for downstream analyses.
```{r load_taxonomy, message=FALSE}
genome_taxonomy <- read_tsv("data/genome_taxonomy.tsv") %>%
  rename(genome = user_genome) %>%
  mutate(genome = str_replace_all(genome,"\\.fa", "")) %>%
  separate(classification, c("domain","phylum","class","order","family","genus","species"),  sep =";") %>%
  select(genome,domain,phylum,class,order,family,genus,species) %>%
  arrange(match(genome, read_counts$genome))
```

#### Genome quality
Quality properties of the genomes. 
```{r load_quality, message=FALSE}
genome_quality <- read_tsv("data/genome_quality.tsv") %>%
  rename(genome = 1) %>%
  mutate(genome = str_replace_all(genome,"\\.fa", "")) %>%
  arrange(match(genome, read_counts$genome)) %>%
  select(genome, Completeness, Contamination, Coding_Density, Genome_Size) %>%
  rename(completeness=Completeness,contamination=Contamination,coding_density=Coding_Density,length=Genome_Size)
```

#### Merged metadata object
Merge taxonomy, length and quality information
```{r create_genomemetadata, message=FALSE}
genome_metadata <- genome_taxonomy %>%
  left_join(genome_quality,by=join_by(genome==genome)) #join quality
```

### Genome tree
This is the raw tree generated by GTDBtk, which needs to be pruned to obtain the phylogenetic tree of the genomes. Note that the archaeal tree is only generated if any archaeans are detected among the genomes.
```{r load_tree, message=FALSE, warning=FALSE}
genome_tree <- read.tree("data/genome_tree.tre")
genome_tree$tip.label <- str_replace_all(genome_tree$tip.label,"'", "") #remove single quotes in MAG names
genome_tree <- keep.tip(genome_tree, tip=read_counts$genome) # keep only MAG tips
```

### MAG functional annotations
This is the raw annotation table generated by DRAM, which is used to generate GIFT data using distillR.
```{r load_annotations, message=FALSE}
genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>%
  rename(gene=1,genome=2)
```

## Filter and normalise data
Raw data needs to be filtered and normalised to make it useful for downstream analyses. 

### Generate coverage table
By dividing the number of base hits by the length of each genome, coverage values can be calculated.

```{r calc_coverage}
genome_coverage <- basehits %>%
  mutate(across(where(is.numeric), ~ ./genome_metadata$length))
```

### Coverage filtering
Genomes that have less than 30% of their length covered by reads are turned into zeros to account for the random allocation of reads across genomes due to mapping heuristics. 

```{r filter_coverage}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]])) 
```

### Generate genome count table
After filtering the low-coverage reads, read counts are transformed into genome counts using genome-length and read-length information.

```{r calc_genometable}
readlength=150 #change if sequencing read length is different
genome_counts_filt <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))

genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

### Distil functional annotations
Raw functional annotations are distilled into genome-inferred functional traits to generate biologically more meaningful functional traits for downstream analyses.

```{r distill_annotations, warning=FALSE, comments="", message=FALSE, results='hide'}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19))
```

## Color scheme

[AlberdiLab](www.alberdilab.dk) projects use unified color schemes developed for the [Earth Hologenome Initiative](www.earthhologenome.org), to facilitate figure interpretation.

```{r get_ehi_colors, warning=FALSE, comments="", message=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)
```

## Wrap working objects
In the last step, the objects that are needed for downstream analyses are stored in an R object.

```{r wrap_objects}
save(read_counts, 
     read_counts_filt, 
     genome_counts, 
     genome_counts_filt, 
     genome_tree, 
     genome_metadata, 
     genome_gifts, 
     sample_metadata, 
     phylum_colors, 
     file = "data/data.Rdata")
```

- **read_counts**: Number of reads mapped to each genome in each sample. Note this is the unfiltered and unnormalised raw community composition table.
- **genome_counts**: Number of genomes quantified in each sample, calculated through filtering and normalising ***read_counts***. This is the community composition table to be used in downstream analyses unless otherwise stated.
- **genome_tree**: Phylogenetic tree of the genomes, to be employed in downstream phylogenetic analyses.
- **genome_metadata**: Taxonomic and quality information of the genomes.
- **sample_metadata**: Treatment/population and other relevant metadata of the samples.

<!--chapter:end:1-data_preparation.Rmd-->

# Data quality

```{r load_data_quality}
load("data/data.Rdata")
```

## Load statistics

```{r load_sequencing_stats, warning=FALSE, comments="", message=FALSE}
read_stats <- read_tsv(c("https://sid.erda.dk/share_redirect/D9oEqxuxql/reports/by_step/reads_data/multiqc_general_stats.txt",
                         "https://sid.erda.dk/share_redirect/e0EsEFnsKu/reports/by_step/reads_data/multiqc_general_stats.txt")) %>%
   mutate(Sample = str_extract(Sample, "M\\d+")) %>%
  rename_with(~str_remove(., "FastQC_mqc-generalstats-fastqc-"), starts_with("FastQC_mqc-generalstats-fastqc-")) %>%
  rename(microsample=Sample) %>%
  select(microsample,percent_duplicates,total_sequences, percent_gc) %>%
  group_by(microsample) %>%
  summarise(total_sequences=sum(total_sequences), percent_duplicates=mean(percent_duplicates), percent_gc=mean(percent_gc))
```

```{r load_host_mapping_stats, warning=FALSE, comments="", message=FALSE}
host_mapping_stats <- read_tsv(c("https://sid.erda.dk/share_redirect/b9ZbTPY0kQ/reports/by_step/preprocess_data/multiqc_samtools_flagstat.txt",
                                 "https://sid.erda.dk/share_redirect/e0EsEFnsKu/reports/by_step/preprocess_data/multiqc_samtools_flagstat.txt")) %>%
    mutate(reference = case_when(
        grepl("GRCh38", Sample, ignore.case = TRUE) ~ "human",
        grepl("GRCg7b", Sample, ignore.case = TRUE) ~ "chicken",
        TRUE ~ NA )) %>%
    filter(reference=="chicken") %>%
    mutate(Sample = str_extract(Sample, "M\\d+")) %>%
    rename(microsample=Sample,reads_mapped_host=mapped_passed,reads_mapped_host_percent=mapped_passed_pct) %>%
    select(microsample,reads_mapped_host,reads_mapped_host_percent) %>%
    group_by(microsample) %>%
    summarise(reads_mapped_host=sum(reads_mapped_host),reads_mapped_host_percent=mean(reads_mapped_host_percent))
```

```{r load_human_mapping_stats, warning=FALSE, comments="", message=FALSE}
human_mapping_stats <- read_tsv(c("https://sid.erda.dk/share_redirect/b9ZbTPY0kQ/reports/by_step/preprocess_data/multiqc_samtools_flagstat.txt",
                                 "https://sid.erda.dk/share_redirect/e0EsEFnsKu/reports/by_step/preprocess_data/multiqc_samtools_flagstat.txt")) %>%
    mutate(reference = case_when(
        grepl("GRCh38", Sample, ignore.case = TRUE) ~ "human",
        grepl("GRCg7b", Sample, ignore.case = TRUE) ~ "chicken",
        TRUE ~ NA )) %>%
    filter(reference=="human") %>%
    mutate(Sample = str_extract(Sample, "M\\d+")) %>%
    rename(microsample=Sample, reads_mapped_human=mapped_passed,reads_mapped_human_percent=mapped_passed_pct) %>%
    select(microsample,reads_mapped_human,reads_mapped_human_percent) %>%
    group_by(microsample) %>%
    summarise(reads_mapped_human=sum(reads_mapped_human),reads_mapped_human_percent=mean(reads_mapped_human_percent))
```

```{r load_mag_mapping_stats, warning=FALSE, comments="", message=FALSE}
quantification_stats  <- read_tsv(c(
  "https://sid.erda.dk/share_redirect/D9oEqxuxql/reports/by_step/quantify_data/multiqc_general_stats.txt",
   "https://sid.erda.dk/share_redirect/e0EsEFnsKu/reports/by_step/quantify_data/multiqc_general_stats.txt")) %>%
   filter(str_detect(Sample, "mgg-pbdrep")) %>%
   mutate(Sample = str_extract(Sample, "M\\d+")) %>%
   rename_with(~str_remove(., "Samtools: stats_mqc-generalstats-samtools_stats-"), 
               starts_with("Samtools: stats_mqc-generalstats-samtools_stats-")) %>%
  rename_with(~str_remove(., "Samtools: flagstat_mqc-generalstats-samtools_flagstat-"),
              starts_with("Samtools: flagstat_mqc-generalstats-samtools_flagstat-")) %>%
   rename(microsample=Sample) %>%
  group_by(microsample) %>%
  summarise(reads_mapped=sum(reads_mapped),reads_mapped_percent=mean(reads_mapped_percent))
```

```{r aggregate_stats, warning=FALSE, comments="", message=FALSE}
quality_stats <- read_stats %>%
    left_join(host_mapping_stats, by=join_by(microsample==microsample)) %>%
    left_join(human_mapping_stats, by=join_by(microsample==microsample)) %>%
    left_join(quantification_stats, by=join_by(microsample==microsample))
```

## Individual overview

### Sequencing depth

```{r sequencing_depth, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=total_sequences,y=microsample,fill=buffer))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454")) +
      geom_vline(xintercept=10000000, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Number of reads", y="Microsamples", fill="Lysis buffer")
```

### Sequence duplication

```{r duplicates, warning=FALSE, comments="", message=FALSE, fig.height=12, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=percent_duplicates,y=microsample,fill=buffer))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454")) +
      geom_vline(xintercept=65, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Percentage of duplicates", y="Microsamples", fill="Lysis buffer")
```

### GC %

```{r gc, warning=FALSE, comments="", message=FALSE, fig.height=12, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=percent_gc,y=microsample,fill=buffer))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454")) +
      geom_vline(xintercept=55, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Percentage of GC", y="Microsamples", fill="Lysis buffer")
```

### Host %

```{r host_percentage, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=reads_mapped_host_percent,y=microsample,fill=buffer))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454")) +
      geom_vline(xintercept=55, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Host %", y="Microsamples", fill="Lysis buffer")
```

### Human %

```{r human_percentage, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=reads_mapped_human_percent,y=microsample,fill=buffer))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454")) +
      geom_vline(xintercept=55, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Human %", y="Microsamples", fill="Lysis buffer")
```

### Bacteria mapping %

```{r bacteria_mapped, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=reads_mapped_percent,y=microsample,fill=buffer))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454")) +
      geom_vline(xintercept=60, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Mapped to MAGs (%)", y="Microsamples", fill="Lysis buffer")
```

## Biplots

### Sequencing depth vs. GC %

```{r sequencing_depth_vs_gc, warning=FALSE, comments="", message=FALSE, fig.height=6, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    filter(type == "Positive") %>%
    ggplot(aes(x=percent_gc,y=total_sequences,color=buffer))+
      geom_point()+
      facet_nested(. ~ batch, scales="free")
```

### Duplicates vs. GC %

```{r duplicates_vs_gc, warning=FALSE, comments="", message=FALSE, fig.height=6, fig.width=10, fig.fullwidth=TRUE}
quality_stats %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    filter(type == "Positive") %>%
    ggplot(aes(x=percent_gc,y=percent_duplicates,color=buffer))+
      geom_point()+
      facet_nested(. ~ batch, scales="free")
```

## Quality flagging

```{r quality_flags, warning=FALSE, comments="", message=FALSE}
quality <- quality_stats %>%
    mutate(depth = case_when(
        total_sequences <= 10000000 ~ 0,
        total_sequences > 10000000 ~ 1,
        TRUE ~ NA)) %>%
    mutate(duplicates = case_when(
        percent_duplicates >= 65 ~ 0,
        percent_duplicates < 65 ~ 1,
        TRUE ~ NA)) %>%
    mutate(gc = case_when(
        percent_gc >= 55 ~ 0,
        percent_gc < 55 ~ 1,
        TRUE ~ NA)) %>%
    mutate(human = case_when(
        reads_mapped_human_percent >= 5 ~ 0,
        reads_mapped_human_percent < 5 ~ 1,
        TRUE ~ NA)) %>%
    mutate(bacteria = case_when(
        reads_mapped_percent <= 60 ~ 0,
        reads_mapped_percent > 60 ~ 1,
        TRUE ~ NA)) %>%
    select(microsample, depth, duplicates, gc, human, bacteria) %>%
    mutate(quality = depth + duplicates + gc + human + bacteria) %>%
    select(microsample, quality)

quality %>% write_tsv("results/quality.tsv")
```

### Quality overview

```{r quality_plot, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
quality %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    ggplot(aes(x=quality,y=microsample,fill=protocol))+
      geom_col()+
      scale_fill_manual(values=c("#6375ad","#d1b454","#d6885e")) +
      geom_vline(xintercept=5, linetype="dashed", color = "red", size=1) + 
      facet_nested(batch + section + type ~ ., scales="free", space="free", switch = "y") +
      theme(strip.text.y.left = element_text(angle = 0)) +
      labs(x="Quality score", y="Microsamples", fill="Library protocol")
```

```{r quality_overview, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
quality %>%
    left_join(sample_metadata,by=join_by(microsample==microsample)) %>%
    filter(section != "Ileum") %>%
    filter(type == "Positive") %>%
    group_by(batch) %>%
    summarise(average=mean(quality, na.rm=TRUE), percentage_5 = mean(quality == 5, na.rm = TRUE) * 100) %>%
    tt()
```

<!--chapter:end:2-data_quality.Rmd-->

# Community composition

```{r load_data_composition}
load("data/data.Rdata")
quality <- read_tsv("results/quality.tsv")
```

## Taxonomy barplot

### Positive samples, coverage-filtered

```{r barplot_positive_filtered, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

genome_counts_filt %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "microsample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(sample_metadata, by = join_by(microsample == microsample)) %>% #append sample metadata
  left_join(quality, by = join_by(microsample == microsample)) %>% #append sample metadata
  mutate(section=unlist(section)) %>%
  filter(!is.na(count)) %>%
  filter(count > 0) %>%
  filter(section != "Ileum") %>%
  filter(type == "Positive") %>%
  filter(quality == 5) %>%
  ggplot(., aes(x=count,y=microsample, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors[-4]) +
    labs(x = "Relative abundance", y="Microsamples") +
    facet_nested(batch + protocol ~ .,  scales="free_y") + #facet per day and treatment
 guides(fill = guide_legend(ncol = 1)) +
    theme(strip.text.y = element_text(angle = 0),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black"),
          panel.spacing = unit(0, "lines")) +
   labs(fill="Phylum")
```

### Positive samples, coverage-unfiltered

```{r barplot_positive_unfiltered, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "microsample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(sample_metadata, by = join_by(microsample == microsample)) %>% #append sample metadata
  left_join(quality, by = join_by(microsample == microsample)) %>% #append sample metadata
  mutate(section=unlist(section)) %>%
  filter(!is.na(count)) %>%
  filter(count > 0) %>%
  filter(section != "Ileum") %>%
  filter(type == "Positive") %>%
  filter(quality == 5) %>%
  ggplot(., aes(x=count,y=microsample, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(x = "Relative abundance", y="Microsamples") +
    facet_nested(batch + protocol ~ .,  scales="free_y") + #facet per day and treatment
 guides(fill = guide_legend(ncol = 1)) +
    theme(strip.text.y = element_text(angle = 0),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black"),
          panel.spacing = unit(0, "lines")) +
   labs(fill="Phylum")

```

### Control samples, coverage-unfiltered

```{r barplot_controls, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    #filter(phylum %in% c("p__Actinomycetota","p__Bacillota","p__Bacillota_A","p__Pseudomonadota","p__Verrucomicrobiota")) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "microsample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(., sample_metadata, by = join_by(microsample == microsample)) %>% #append sample metadata
  filter(is.na(Xcoord)) %>%
  filter(type %in% c("NegativeMembrane","NegativeCollection","NegativeReaction")) %>%
  ggplot(., aes(x=count,y=microsample, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(x = "Relative abundance", y="Membrane controls") +
    facet_nested(batch + type + protocol ~ .,  scales="free_y") + #facet per day and treatment
 guides(fill = guide_legend(ncol = 1)) +
    theme(strip.text.y = element_text(angle = 0),
          axis.text.y = element_blank(),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black"),
          panel.spacing = unit(0, "lines")) +
   labs(fill="Phylum")
```

```{r heatmap_positive_filtered, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
vertical_tree <- force.ultrametric(genome_tree,method="extend") %>%
        ggtree(., size = 0.3) + geom_tiplab()

phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    mutate(phylum = factor(phylum, levels = unique(phylum))) %>%
    column_to_rownames(var = "genome") %>%
    select(phylum)

colors_alphabetic <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
  select(phylum, colors) %>%
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

vertical_tree <- gheatmap(vertical_tree, phylum_colors, offset=-0.3, width=0.1, colnames=FALSE) +
    scale_fill_manual(values=colors_alphabetic) +
    new_scale_fill()

sample_selection <- sample_metadata %>%
      filter(!is.na(Xcoord)) %>%
      filter(cryosection == "G121eI101B") %>%
      left_join(quality, by=join_by(microsample==microsample)) %>%
      filter(quality>=5) %>% select(microsample) %>% pull()

genome_counts_selected <- genome_counts_filt %>%
          select(all_of(c("genome",sample_selection))) %>% column_to_rownames(var="genome") %>% tss()

vertical_tree <- gheatmap(vertical_tree, genome_counts_selected, offset=-0.2, width=0.5, colnames=FALSE, colnames_angle=90, font.size=3, colnames_position="top", colnames_offset_y = 15) +
    vexpand(.08) +
    coord_cartesian(clip = "off") +
    scale_fill_gradient(low = "#f4f4f4", high = "#315b7d", na.value="#f4f4f4") +
    new_scale_fill()

vertical_tree
```

<!--chapter:end:3-community_composition.Rmd-->

# Alpha diversity

## Filtered

```{r alpha_diversities_filtered, warning=FALSE, comments="", message=FALSE}
#Calculate Hill numbers
richness <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(where(~!all(. == 0))) %>%
            hilldiv(.,q=0) %>%
            t() %>%
            as.data.frame() %>%
            rename(richness=1) %>%
            rownames_to_column(var="microsample")

neutral <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(where(~!all(. == 0))) %>%
            hilldiv(.,q=1) %>%
            t() %>%
            as.data.frame() %>%
            rename(neutral=1) %>%
            rownames_to_column(var="microsample")

phylogenetic <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(where(~!all(. == 0))) %>%
            hilldiv(.,q=1,tree=genome_tree) %>%
            t() %>%
            as.data.frame() %>%
            rename(phylogenetic=1) %>%
            rownames_to_column(var="microsample")

# Merge alpha diversities
alpha_diversity <- richness %>%
      full_join(neutral,by=join_by(microsample==microsample)) %>%
      full_join(phylogenetic,by=join_by(microsample==microsample)) %>%
      left_join(sample_metadata, by=join_by(microsample==microsample))

# Write alpha diversities
alpha_diversity %>% write_tsv("results/alpha_div_filtered.tsv")

# Print alpha diversity
alpha_diversity %>%
  select(microsample,richness, neutral, phylogenetic, section, buffer, batch, protocol, Xcoord, Ycoord, size) %>%
  tt()
  
# Print alpha diversity summary
alpha_diversity %>%
  select(richness, neutral, phylogenetic, section, batch, protocol) %>%
  group_by(section, batch, protocol) %>%
  summarise(richness=mean(richness), neutral=mean(neutral), phylogenetic=mean(phylogenetic)) %>%
  tt()
```

```{r alpha_diversities_filtered_plot, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
alpha_diversity %>%
  pivot_longer(!c(microsample,section,cryosection,type,buffer,batch,Xcoord,Ycoord,size), names_to = "metric", values_to = "value") %>%
  left_join(quality,by=join_by(microsample==microsample)) %>%
  mutate(metric=factor(metric,levels=c("richness","neutral","phylogenetic"))) %>%
  ggplot(aes(x=batch, y=value, color=quality, group=batch))+ 
        scale_color_gradient(low = "red", high = "green", name = "Quality") +
        geom_boxplot(outlier.shape = NA) +
        geom_jitter(width=0.2) + 
        facet_grid(metric ~ section, scales = "free")
```

## Unfiltered

```{r alpha_diversities_unfiltered, warning=FALSE, comments="", message=FALSE}
#Calculate Hill numbers
richness <- genome_counts %>%
            column_to_rownames(var="genome") %>%
            select(where(~!all(. == 0))) %>%
            hilldiv(.,q=0) %>%
            t() %>%
            as.data.frame() %>%
            rename(richness=1) %>%
            rownames_to_column(var="microsample")

neutral <- genome_counts %>%
            column_to_rownames(var="genome") %>%
            select(where(~!all(. == 0))) %>%
            hilldiv(.,q=1) %>%
            t() %>%
            as.data.frame() %>%
            rename(neutral=1) %>%
            rownames_to_column(var="microsample")

phylogenetic <- genome_counts %>%
            column_to_rownames(var="genome") %>%
            select(where(~!all(. == 0))) %>%
            hilldiv(.,q=1,tree=genome_tree) %>%
            t() %>%
            as.data.frame() %>%
            rename(phylogenetic=1) %>%
            rownames_to_column(var="microsample")

# Merge alpha diversities
alpha_diversity <- richness %>%
      full_join(neutral,by=join_by(microsample==microsample)) %>%
      full_join(phylogenetic,by=join_by(microsample==microsample)) %>%
      left_join(sample_metadata, by=join_by(microsample==microsample))

# Write alpha diversities
alpha_diversity %>% write_tsv("results/alpha_div_unfiltered.tsv")

# Print alpha diversity
alpha_diversity %>%
  select(microsample,richness, neutral, phylogenetic, section, buffer, Xcoord, Ycoord, size) %>%
  tt()
  
# Print alpha diversity summary
alpha_diversity %>%
  select(richness, neutral, phylogenetic, section, buffer) %>%
  group_by(section, buffer) %>%
  summarise(richness=mean(richness), neutral=mean(neutral), phylogenetic=mean(phylogenetic)) %>%
  tt()
```

```{r alpha_diversities_unfiltered_plot, warning=FALSE, comments="", message=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
alpha_diversity %>%
  pivot_longer(!c(microsample,section,cryosection,type,buffer,batch,Xcoord,Ycoord,size), names_to = "metric", values_to = "value") %>%
  left_join(quality,by=join_by(microsample==microsample)) %>%
  mutate(metric=factor(metric,levels=c("richness","neutral","phylogenetic"))) %>%
  ggplot(aes(x=batch, y=value, color=quality, group=batch))+ 
        scale_color_gradient(low = "red", high = "green", name = "Quality") +
        geom_boxplot(outlier.shape = NA) +
        geom_jitter(width=0.2) + 
        facet_grid(metric ~ section, scales = "free")
```

<!--chapter:end:4_alpha_diversity.Rmd-->

# Beta diversity

## Caecum

```{r beta_diversities_caecum, warning=FALSE, comments="", message=FALSE}
caecum_samples <- sample_metadata %>% filter(cryosection == "G121eI101B") %>% filter(!is.na(Xcoord)) %>% select(microsample) %>% pull()

#Calculate Hill numbers
richness_caecum <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(all_of(caecum_samples)) %>%
            select(where(~!all(. == 0))) %>%
            hillpair(.,q=0, metric="C", out="pair")

neutral_caecum <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(all_of(caecum_samples)) %>%
            select(where(~!all(. == 0))) %>%
            hillpair(.,q=1, metric="C", out="pair")

phylogenetic_caecum <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(all_of(caecum_samples)) %>%
            select(where(~!all(. == 0))) %>%
            hillpair(.,q=1, tree=genome_tree, metric="C", out="pair")

# Merge beta diversities
beta_diversity <- richness_caecum %>%
      full_join(neutral_caecum,by=c("first", "second")) %>%
      full_join(phylogenetic_caecum,by=c("first", "second")) %>%
      rename(richness=C.x, neutral=C.y, phylogenetic=C)

# Write alpha diversities
beta_diversity %>% write_tsv("results/beta_div_caecum.tsv")
```

```{r beta_diversities_caecum_plot, fig.height=8, fig.width=10, fig.fullwidth=TRUE, eval=FALSE}

#Select high-quality samples
caecum_samples_hq <- sample_metadata %>% 
  left_join(quality,by=join_by(microsample==microsample)) %>%
  filter(section != "Ileum") %>% 
  filter(quality >= 5) %>% 
  filter(!is.na(Xcoord)) %>% 
  select(microsample) %>% 
  pull()

caecum_samples_nmds <- genome_counts_filt %>%
            column_to_rownames(var="genome") %>%
            select(all_of(caecum_samples_hq)) %>%
            select(where(~!all(. == 0))) %>%
            hillpair(.,q=1, metric="C", out="dist") %>%
            metaMDS(.,trymax = 999, k=2, trace=0) %>%
            vegan::scores() %>%
            as_tibble(., rownames = "microsample") %>%
            left_join(sample_metadata, by = join_by(microsample == microsample))

caecum_samples_nmds %>%
            ggplot(., aes(x=NMDS1,y=NMDS2, color=protocol, shape=batch)) +
                geom_point(size=2) +
                scale_color_manual(values=c("#6375ad","#d1b454","#d6885e")) +
                theme_classic() +
                theme(legend.position="right", legend.box="vertical") +
                guides(color=guide_legend(title="Region"))

```

<!--chapter:end:5-beta_diversity.Rmd-->

